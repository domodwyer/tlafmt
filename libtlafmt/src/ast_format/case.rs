use tree_sitter::Node;

use crate::{format_node, token::Token, EmptyLines, Error, Renderer};

/// Format a CASE block for `def`, indenting the arms and ensuring each arm is
/// spaced by a single newline.
pub(super) fn format_case<'a, W>(
    def: Node<'a>,
    input: &'a str,
    empty_lines: &mut EmptyLines,
    writer: &mut Renderer<'a, W>,
) -> Result<(), Error>
where
    W: std::io::Write,
{
    let mut c = def.walk();
    let mut iter = def.named_children(&mut c).peekable();

    empty_lines.maybe_insert(&def, writer)?;
    writer.indent_inc();

    writer.push(Token::KeywordCase)?;

    // True when the nodes being visited are part of the initial CASE
    // <condition> and false once the first case_box is observed.
    let mut in_condition_expr = true;

    while let Some(n) = iter.next() {
        match n.kind() {
            "case_box" => {
                writer.push(Token::Newline)?;
                writer.push(Token::CaseBox)?;
            }
            "other_arm" | "case_arm" => {
                empty_lines.suppress(&n);
                if let Some(next) = iter.peek() {
                    empty_lines.suppress(next);
                }

                writer.indent_inc();
                format_node(n, input, empty_lines, writer)?;
                writer.indent_dec();

                if in_condition_expr {
                    writer.indent_inc();
                    in_condition_expr = false;
                }
            }
            _ => format_node(n, input, empty_lines, writer)?,
        }
    }

    // Once for the CASE, once for the arms.
    writer.indent_dec();
    writer.indent_dec();

    Ok(())
}

#[cfg(test)]
mod tests {
    use crate::assert_rewrite;

    #[test]
    fn test_typical() {
        assert_rewrite!(
            r#"
---- MODULE B ----
X == CASE self \in RM -> "RS"
            [] self = 0 -> "TS"
            [] self = 10 -> "BTS"
====
"#
        );
    }

    #[test]
    fn test_comments() {
        assert_rewrite!(
            r#"
---- MODULE B ----
X == CASE self \in RM -> "RS"
            \* Something
            [] self = 0 -> "TS"
            (* bananas *)
            [] self = 10 -> "BTS" \* platanos
            [] self = 10 -> "BTS"
====
"#
        );
    }

    #[test]
    fn test_multiline_pointer() {
        assert_rewrite!(
            r#"
---- MODULE B ----
X == CASE self \in RM -> "RS"
            [] self = 0 -> "TS"
            [] self = 10 ->
                 "BTS"
====
"#
        );
    }

    #[test]
    fn test_conj_arms() {
        assert_rewrite!(
            r#"
---- MODULE B ----
X == CASE self \in RM -> "RS"
            [] /\ self = 0
               /\ self = 0 -> "TS"
            [] self = 10 -> "BTS" /\ A = 42
====
"#
        );
    }

    #[test]
    fn test_keyword_line() {
        assert_rewrite!(
            r#"
---- MODULE B ----
X == CASE
self \in RM -> "RS"
        [] self = 0 -> "TS"
            [] self = 10 -> "BTS"
====
"#
        );
    }

    #[test]
    fn test_exploded() {
        assert_rewrite!(
            r#"
---- MODULE B ----
X ==
CASE
self \in RM -> "RS"
[] self = 0 -> "TS"
[] self = 10 -> "BTS"
====
"#
        );
    }

    #[test]
    fn test_box_arm_newlines() {
        assert_rewrite!(
            r#"
---- MODULE B ----
X ==
CASE
self \in RM -> "RS"
[]
        self = 0 -> "TS"

        []

self = 10 -> "BTS"
====
"#
        );
    }

    #[test]
    fn test_other() {
        assert_rewrite!(
            r#"
---- MODULE B ----
X == CASE \/ x = 0 \/ y = 0
          \/ x > N \/ y > N
          \/ ~grid[<<x, y>>] -> 0
        [] OTHER -> 1
====
"#
        );
    }

    /// Discovered by fuzzing, this test reproduces a CASE node that contains an
    /// ERROR child node.
    #[test]
    fn test_fuzz_error_node_kind() {
        assert_rewrite!(
            r#"
---- MODULE B ----
==CASE>d->0
====
"#
        );
    }
}
