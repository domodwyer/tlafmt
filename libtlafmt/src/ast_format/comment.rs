use tree_sitter::Node;

use crate::{
    get_str,
    helpers::Indent,
    token::{Position, Token},
    Error, Renderer,
};

/// Emit a [`Token::Comment`] for `def`, after processing to adjust a comment
/// that may be attributed to an indented list that is actually adjacent to the
/// next AST node.
pub(super) fn format_comment<'a, W>(
    def: Node<'_>,
    input: &'a str,
    writer: &mut Renderer<'a, W>,
) -> Result<(), Error>
where
    W: std::io::Write,
{
    // Block comments should be rendered "as-is" without additional indentation.
    //
    // If a comment was not indented, it should be rendered without
    // formatter-added indentation below this branch.
    if def.start_position().column != 0 {
        // Special case indentation within an operator body to ensure a comment
        // appears indented if it is the first statement in the body.
        if def
            .parent()
            .is_some_and(|v| v.kind() == "operator_definition")
        {
            let orig = writer.indent_get();
            writer.indent_set(std::cmp::max(Indent::new(1), orig));
            writer.push(Token::Comment(get_str(&def, input), Position::from(&def)))?;
            writer.indent_set(orig);
        } else {
            writer.push(Token::Comment(get_str(&def, input), Position::from(&def)))?;
        }

        return Ok(());
    }

    let orig = writer.indent_get();

    writer.indent_set(Indent::new(0));
    let ret = writer.push(Token::Comment(get_str(&def, input), Position::from(&def)));
    writer.indent_set(orig);

    ret.map_err(Into::into)
}

#[cfg(test)]
mod tests {
    use crate::assert_rewrite;

    #[test]
    fn test_comment_let_in_list() {
        assert_rewrite!(
            r"
---- MODULE Bananas ------
A ==
    LET B == 42
    IN /\ C = D

\* Where is this placed?
B == 24
=============================================================================
"
        );
    }

    #[test]
    fn test_comment_with_lists() {
        assert_rewrite!(
            r"
---- MODULE Bananas ------
DoStuff ==
    /\ A' = 1
    \* Comment in a list.
    /\ B' = 2

\* Some comment for the operator.
Another == 42
====="
        );
    }

    #[test]
    fn test_operator_long_form_inline_comment() {
        assert_rewrite!(
            "\
---- MODULE Bananas ------
DoStuff(
b,
a  , \\* platanos?
n,

A,		N     ,
a,
S (* are great
    dont
    you think*)
) == 42
====="
        );
    }

    #[test]
    fn test_block() {
        assert_rewrite!(
            r"
---- MODULE Bananas ------
X ==
    /\ A = 42
    (* There's a block comment here *)
    /\ B = A
====="
        );
    }
}
