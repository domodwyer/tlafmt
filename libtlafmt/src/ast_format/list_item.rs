use tree_sitter::Node;

use crate::{format_node, token::Token, EmptyLines, Error, Renderer};

/// Render a conjunctive or disjunctive list item, indenting the body of the
/// item by 1.
pub(super) fn format_list_item<'a, W>(
    def: Node<'a>,
    input: &'a str,
    empty_lines: &mut EmptyLines,
    writer: &mut Renderer<'a, W>,
) -> Result<(), Error>
where
    W: std::io::Write,
{
    empty_lines.maybe_insert(&def, writer)?;
    writer.push(Token::Newline)?;

    let mut c = def.walk();
    let iter = def.named_children(&mut c).peekable();
    for n in iter {
        match n.kind() {
            "bullet_conj" => {
                writer.push(Token::And)?;
                writer.indent_inc();
            }
            "bullet_disj" => {
                writer.push(Token::Or)?;
                writer.indent_inc();
            }
            _ => format_node(n, input, empty_lines, writer)?,
        }
    }

    writer.indent_dec();

    Ok(())
}
