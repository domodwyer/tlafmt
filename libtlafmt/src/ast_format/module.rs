use std::{io::Write, iter::Peekable};

use tree_sitter::Node;

use crate::{ast_format::format_node, get_str, helpers::EmptyLines, token::Token, Error, Renderer};

/// Format a TLA module.
pub(super) fn format_module<'a, 'b: 'a, W>(
    n: Node<'b>,
    input: &'a str,
    empty_lines: &mut EmptyLines,
    mut out: &mut Renderer<'a, W>,
) -> Result<(), Error>
where
    W: Write,
{
    assert_eq!(n.kind(), "module"); // Validated by caller.

    let mut c = n.walk();
    let mut iter = c.node().named_children(&mut c).peekable();

    while let Some(node) = iter.peek() {
        // Emit newlines, squashing repeated newlines to at most 1 empty line.
        empty_lines.maybe_insert(node, &mut out)?;

        match node.kind() {
            "header_line" => format_module_header(&mut iter, input, &mut out),
            _ => format_node(iter.next().unwrap(), input, empty_lines, out),
        }?;
    }

    out.flush()?;
    Ok(())
}

/// Consume the module header nodes from `iter`, printing a normalised module
/// header.
fn format_module_header<'a, W>(
    iter: &mut Peekable<impl ExactSizeIterator<Item = Node<'a>>>,
    input: &'a str,
    out: &mut Renderer<'a, W>,
) -> Result<(), Error>
where
    W: Write,
{
    // Greedily try and consume the header line, which is composed of 3 nodes:
    //
    //   * header_line
    //   * identifier
    //   * header_line
    //
    // All must be consumed from the iterator.
    let left = iter.next().unwrap();
    assert_eq!(left.kind(), "header_line"); // Validated by caller.

    let ident = match iter.next_if(|v| v.kind() == "identifier") {
        Some(v) => v,
        None => return Err(Error::ModuleHeader),
    };

    match iter.next_if(|v| v.kind() == "header_line") {
        Some(_) => {}
        None => return Err(Error::ModuleHeader),
    };

    let name = get_str(&ident, input).trim_ascii();

    out.push(Token::ModuleHeader(name))?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use crate::assert_rewrite;

    #[test]
    fn test_module_header_length_normalisation() {
        assert_rewrite!(
            "\
Some bananas
---- MODULE DifferentialEquations ------
LOCAL X == 42
====="
        );
    }

    #[test]
    fn test_module_header_whitespace_trim() {
        assert_rewrite!(
            "\
     ----    MODULE     DifferentialEquations   ------
====="
        );
    }

    /// Reduce overly long module lines to fit into the desired line width.
    #[test]
    fn test_module_header_length_trim() {
        assert_rewrite!(
            "\
---------------------------------- MODULE DifferentialEquations ----------------------------------
====="
        );
    }

    #[test]
    fn test_module_header_even_odd_length() {
        // Line is perfectly even.
        assert_rewrite!(
            "\
--------------------------------- MODULE A ---------------------------------
================================================================================"
        );
        // Text is odd and equal dashes would result in an excessively long line.
        assert_rewrite!(
            "\
--------------------------------- MODULE AB ---------------------------------
================================================================================"
        );
    }

    #[test]
    fn test_newline_squashing() {
        assert_rewrite!(
            "\
----------------------------------- MODULE A -----------------------------------
LOCAL C == 42
LOCAL A == 42



\\* Comment

LOCAL T == 13
(* Another *)




(* BANANAS *)
LOCAL S == 24


================================================================================"
        );
    }
}
