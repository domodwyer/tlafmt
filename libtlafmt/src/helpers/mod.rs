mod indent;
mod newline;

pub(crate) use indent::*;
pub(crate) use newline::*;
