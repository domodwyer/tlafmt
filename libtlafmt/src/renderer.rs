use std::io::Write;

use crate::{helpers::IndentDecorator, token::Token, LINE_WIDTH};

/// A renderer of [`Token`] instances, writing the resulting output to `W`.
pub(crate) struct Renderer<'a, W> {
    /// The current indentation depth.
    indent_depth: usize,

    /// A decorator over `W` that emits indentation after every rendered
    /// newline.
    indent: IndentDecorator<W>,

    /// A buffer containing the lowered formatter [`Token`] to write to `W`.
    buf: Vec<(Token<'a>, usize)>,

    /// True when the last token wrote to `ident` was a newline.
    last_token_was_newline: bool,
}

impl<'a, W> Renderer<'a, W>
where
    W: std::io::Write,
{
    /// Initialise a [`Renderer`] to write to `out`.
    pub(crate) fn new(out: W) -> Self {
        Self {
            indent_depth: 0,
            indent: IndentDecorator::new(out),
            buf: Default::default(),
            last_token_was_newline: false,
        }
    }

    /// Read the current indentation depth.
    pub(crate) fn indent_get(&self) -> usize {
        self.indent_depth
    }

    /// Increase the indentation depth.
    pub(crate) fn indent_inc(&mut self) {
        self.indent_depth += 1;
    }

    /// Decrease the indentation depth.
    pub(crate) fn indent_set(&mut self, v: usize) {
        self.indent_depth = v;
    }

    /// Decrement the indentation depth.
    ///
    /// # Panics
    ///
    /// Panics if the indentation depth is 0.
    pub(crate) fn indent_dec(&mut self) {
        debug_assert_ne!(self.indent_depth, 0);

        self.indent_depth -= 1;
    }

    /// Add `t` to the render queue.
    pub(crate) fn push(&mut self, t: Token<'a>) -> Result<(), std::io::Error> {
        // Incrementally emit the tokens - this is not required for correctness.
        if matches!(t, Token::SourceNewline | Token::Newline) {
            self.buf.push((t, self.indent_depth));
            return self.flush();
        }

        self.buf.push((t, self.indent_depth));

        Ok(())
    }

    /// Flush the queue of [`Token`], rendering them to the output sink.
    pub(crate) fn flush(&mut self) -> Result<(), std::io::Error> {
        let mut iter = self.buf.drain(..).peekable();

        while let Some((t, indent_depth)) = iter.next() {
            self.indent.set(indent_depth);

            // If this token cannot appear before the next token, skip rendering
            // this one.
            if let Some((next, _)) = iter.peek() {
                if !t.can_precede(next) {
                    continue;
                }
            }

            let s = match &t {
                Token::StepOrStutter(ident) => {
                    self.indent.write_all(format!("[{ident}]_").as_bytes())?;
                    continue;
                }

                Token::Newline if self.last_token_was_newline => {
                    // Prevent the formatter from unintentionally inserting a
                    // forced newline in addition to a newline that was in the
                    // source spec.
                    continue;
                }

                Token::ModuleHeader(name) => {
                    const MODULE: &str = " MODULE ";
                    let line_len = LINE_WIDTH
                        .checked_sub(name.len())
                        .and_then(|v| v.checked_sub(MODULE.len() + 1))
                        .and_then(|v| v.checked_div(2))
                        .unwrap_or(1);

                    // Sometimes the module dashes can't be exactly equal and fill up the entire
                    // line because the "MODULE" and name have an odd length.
                    //
                    // When this happens, pad the right-side line with an extra dash.
                    let mut right_extra = 0;
                    if (line_len * 2) + MODULE.len() + 1 + name.len() == LINE_WIDTH - 1 {
                        right_extra = 1;
                    }

                    write!(
                        &mut self.indent,
                        "{}{MODULE}{name} {}",
                        "-".repeat(line_len),
                        "-".repeat(line_len + right_extra)
                    )?;

                    continue;
                }

                Token::LineDivider(c) => {
                    let s = std::iter::repeat(c).take(LINE_WIDTH).collect::<String>();
                    self.indent.write_all(s.as_bytes())?;
                    continue;
                }

                Token::Raw(s) | Token::Comment(s) => s,
                Token::Prime => "'",
                Token::Always => "[]",
                Token::Eventually => "<>",
                Token::Newline | Token::SourceNewline => "\n",
                Token::KeywordChoose => "CHOOSE",
                Token::KeywordLet => "LET",
                Token::KeywordIn => "IN",
                Token::KeywordLocal => "LOCAL",
                Token::KeywordInstance => "INSTANCE",
                Token::KeywordDomain => "DOMAIN",
                Token::KeywordIf => "IF",
                Token::KeywordThen => "THEN",
                Token::KeywordElse => "ELSE",
                Token::KeywordExtends => "EXTENDS",
                Token::KeywordConstant => "CONSTANT",
                Token::KeywordConstants => "CONSTANTS",
                Token::KeywordVariable => "VARIABLE",
                Token::KeywordVariables => "VARIABLES",
                Token::KeywordExcept => "EXCEPT",
                Token::KeywordUnchanged => "UNCHANGED",
                Token::KeywordEnabled => "ENABLED",
                Token::KeywordSubset => "SUBSET",
                Token::MapTo => ":>",
                Token::MapsTo => "->",
                Token::AllMapsTo => "|->",
                Token::Compose => "@@",
                Token::Exists => r"\E",
                Token::All => r"\A",
                Token::SetIn => r"\in",
                Token::SetNotIn => r"\notin",
                Token::And => r"/\",
                Token::Or => r"\/",
                Token::Not => "~",
                Token::Ident(s) => s,
                Token::Lit(s) => s,
                Token::ParenOpen => "(",
                Token::ParenClose => ")",
                Token::SquareOpen => "[",
                Token::SquareClose => "]",
                Token::CurlyOpen => "{",
                Token::CurlyClose => "}",
                Token::AngleOpen => "<<",
                Token::AngleClose => ">>",
                Token::Comma => ",",
                Token::SemiColon => ":",
                Token::Bang => "!",
                Token::At => "@",
                Token::Eq => "=",
                Token::Eq2 => "==",
                Token::NotEq => "/=",
                Token::SubsetEq => r"\subseteq",
                Token::Dot => ".",
                Token::Dots2 => "..",
                Token::Plus => "+",
                Token::Minus => "-",
                Token::Multiply => "*",
                Token::AppendShort => r"\o",
                Token::Real => r"Real",
                Token::GreaterThan => ">",
                Token::GreaterThanEqual => ">=",
                Token::LessThan => "<",
                Token::LessThanEqual => "<=",
                Token::SetMinus => r"\",
                Token::Divide => r"/",
                Token::True => "TRUE",
                Token::False => "FALSE",
                Token::Union => r"\union",
                Token::Intersect => r"\intersect",
                Token::WeakFairness => "WF_",
                Token::StrongFairness => "SF_",
                Token::Implies => "=>",
                Token::KeywordTheorem => "THEOREM",
            };

            // Write the rendered token.
            self.indent.write_all(s.as_bytes())?;

            // Insert a space if this node and the next node can be space
            // delimited.
            if iter.peek().is_some_and(|(v, _)| t.is_space_delimited(v)) {
                self.indent.write_all(b" ")?;
            }

            self.last_token_was_newline = matches!(t, Token::SourceNewline | Token::Newline);
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn format<'a>(tokens: impl IntoIterator<Item = Token<'a>>) -> String {
        let mut buf = Vec::new();
        let mut w = Renderer::new(&mut buf);

        for t in tokens {
            w.push(t).unwrap();
        }

        w.flush().unwrap();
        drop(w);

        String::from_utf8(buf).expect("valid utf8 output")
    }

    #[test]
    fn test_write_line() {
        let output = format([Token::Raw("testing")]);
        assert_eq!(output, "testing");
    }

    #[test]
    fn test_liveness() {
        let output = format([
            Token::Eq2,
            Token::Eventually,
            Token::Always,
            Token::ParenOpen,
            Token::Ident("bananas"),
            Token::ParenClose,
        ]);
        assert_eq!(output, "== <>[](bananas)");
    }

    #[test]
    fn test_spec_next() {
        let output = format([
            Token::And,
            Token::Always,
            Token::StepOrStutter("Next"),
            Token::Ident("vars"),
        ]);
        assert_eq!(output, r"/\ [][Next]_vars");
    }

    #[test]
    fn test_range() {
        let output = format([Token::Ident("n"), Token::Dots2, Token::Ident("m")]);
        assert_eq!(output, "n..m");

        let output = format([
            Token::Ident("n"),
            Token::Dots2,
            Token::ParenOpen,
            Token::Ident("m"),
            Token::Plus,
            Token::Lit("1"),
            Token::ParenClose,
        ]);
        assert_eq!(output, "n..(m + 1)");
    }

    #[test]
    fn test_op_params() {
        let output = format([
            Token::Ident("bananas"),
            Token::ParenOpen,
            Token::Ident("A"),
            Token::Comma,
            Token::Ident("B"),
            Token::Comma,
            Token::Ident("C"),
            Token::Comma,
            Token::ParenClose,
        ]);
        assert_eq!(output, "bananas(A, B, C)");
    }

    #[test]
    fn test_fn_application() {
        let output = format([
            Token::Ident("bananas"),
            Token::SquareOpen,
            Token::Ident("platanos"),
            Token::SquareClose,
            Token::SquareOpen,
            Token::Lit("42"),
            Token::SquareClose,
        ]);
        assert_eq!(output, "bananas[platanos][42]");
    }

    #[test]
    fn test_record_fields() {
        let output = format([
            Token::Ident("bananas"),
            Token::Dot,
            Token::Ident("platanos"),
        ]);
        assert_eq!(output, "bananas.platanos");
    }

    #[test]
    fn test_except_bang() {
        let output = format([
            Token::KeywordExcept,
            Token::Bang,
            Token::Dot,
            Token::Ident("bananas"),
        ]);
        assert_eq!(output, "EXCEPT !.bananas");

        let output = format([
            Token::KeywordExcept,
            Token::Bang,
            Token::SquareOpen,
            Token::Ident("bananas"),
            Token::SquareClose,
        ]);
        assert_eq!(output, "EXCEPT ![bananas]");
    }

    #[test]
    fn test_gt_lt_parens() {
        let input = [
            (Token::GreaterThan, ">"),
            (Token::GreaterThanEqual, ">="),
            (Token::LessThan, "<"),
            (Token::LessThanEqual, "<="),
        ];

        for (token, symbol) in input {
            let output = format([
                Token::Ident("bananas"),
                token,
                Token::ParenOpen,
                Token::Ident("n"),
                Token::Plus,
                Token::Lit("1"),
                Token::ParenClose,
            ]);
            assert_eq!(output, format!("bananas {symbol} (n + 1)"));
        }
    }

    #[test]
    fn test_not() {
        let output = format([Token::Not, Token::Ident("bananas")]);
        assert_eq!(output, "~bananas");
    }

    #[test]
    fn test_fairness() {
        let output = format([Token::WeakFairness, Token::Ident("vars")]);
        assert_eq!(output, "WF_vars");

        let output = format([Token::StrongFairness, Token::Ident("vars")]);
        assert_eq!(output, "SF_vars");
    }

    #[test]
    fn test_stutter() {
        let output = format([Token::StepOrStutter("Next"), Token::Ident("vars")]);
        assert_eq!(output, "[Next]_vars");
    }

    #[test]
    fn test_bounded_quantification() {
        let output: String = format([
            Token::Exists,
            Token::Ident("t"),
            Token::SetIn,
            Token::Ident("vars"),
            Token::SemiColon,
            Token::Ident("bananas"),
        ]);
        assert_eq!(output, r"\E t \in vars: bananas");
    }

    #[test]
    fn test_prime_var() {
        let output: String = format([
            Token::Ident("bananas"),
            Token::Prime,
            Token::Eq,
            Token::Lit("42"),
        ]);
        assert_eq!(output, "bananas' = 42");
    }

    #[test]
    fn test_eq_paren() {
        let parens = [
            (Token::ParenOpen, "("),
            (Token::SquareOpen, "["),
            (Token::AngleOpen, "<< "), // Extra space
        ];

        let eq = [(Token::Eq, "="), (Token::NotEq, "/=")];

        for (paren, p_symbol) in parens {
            for (eq, e_symbol) in &eq {
                let output: String = format([
                    eq.clone(),
                    paren.clone(),
                    Token::KeywordEnabled,
                    Token::Ident("Bananas"),
                ]);

                assert_eq!(output, format!("{e_symbol} {p_symbol}ENABLED Bananas"));
            }
        }
    }

    #[test]
    fn test_compose_paren() {
        let output: String = format([
            Token::Ident("bananas"),
            Token::Compose,
            Token::ParenOpen,
            Token::Ident("x"),
            Token::MapTo,
            Token::Lit("42"),
            Token::ParenClose,
        ]);
        assert_eq!(output, "bananas @@ (x :> 42)");
    }

    /// This test covers the unhandled tokens in the AST that are emitted as-is.
    /// Spacing is a best guess in this case, as the original may or may not
    /// have contained spacing between or after the raw token.
    #[test]
    fn test_raw_tokens() {
        let output: String = format([Token::Ident("bananas"), Token::Raw("!!!"), Token::Lit("42")]);
        assert_eq!(output, "bananas !!! 42");
    }
}
