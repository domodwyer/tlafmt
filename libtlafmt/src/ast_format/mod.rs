//! Lowering methods that emit formatting [`Token`] from AST [`Node`].
//!
//! [`Token`]: crate::token::Token
//! [`Node`]: tree_sitter::Node

mod comment;
mod module;
mod node;

use comment::*;
use module::*;
pub(crate) use node::*;
