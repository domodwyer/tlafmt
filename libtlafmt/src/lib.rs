#![doc = include_str!("../README.md")]

//   Copyright 2025 Dom Dwyer <dom@itsallbroken.com>
//
//   Licensed under the Apache License, Version 2.0 (the "License");
//   you may not use this file except in compliance with the License.
//   You may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS,
//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//   See the License for the specific language governing permissions and
//   limitations under the License.

mod ast_format;
mod helpers;
mod renderer;
mod test_utils;
mod token;

use std::{borrow::Cow, io::Write};

use ast_format::format_node;
use helpers::{EmptyLines, INDENT_STR};
use renderer::Renderer;
use thiserror::Error;
use tree_sitter::{Node, Parser, Tree};

const LINE_WIDTH: usize = 80;

/// Errors during AST parsing, lowering or rendering.
#[derive(Debug, Error)]
pub enum Error {
    /// An I/O error writing to the output sink.
    #[error("data store disconnected")]
    IO(#[from] std::io::Error),

    /// The input cannot be parsed into an AST.
    #[error("unknown parser error")]
    Parse,

    /// The module header is malformed.
    #[error("invalid module header")]
    ModuleHeader,

    /// A `[Next]_vars` sequence is malformed.
    #[error("invalid step-or-stutter sequence")]
    StepOrStutter,
}

/// A parsed TLA file ready for formatting.
#[derive(Debug)]
pub struct ParsedFile<'a> {
    t: Tree,
    input: Cow<'a, str>,
}

impl<'a> ParsedFile<'a> {
    /// Parse the `input` TLA spec into an AST.
    pub fn new(input: &'a str) -> Result<Self, Error> {
        let mut parser = Parser::new();
        parser
            .set_language(&tree_sitter_tlaplus::LANGUAGE.into())
            .expect("error loading TLA+ grammar");

        // Normalise tab characters - if a mixture of tab and space are used,
        // the AST may produce incorrect nodes. Specifically conj_items using
        // tabs can become a bound_infix_op instead of conj_list, see
        // `test_mixed_tabs_spaces`.
        let input = match input.contains("\t") {
            true => Cow::Owned(input.replace("\t", INDENT_STR)),
            false => Cow::Borrowed(input),
        };

        Ok(Self {
            t: parser.parse(input.as_bytes(), None).ok_or(Error::Parse)?,
            input,
        })
    }

    /// Format and render the parsed spec into `out`.
    ///
    /// # Errors
    ///
    /// If formatting fails `out` may contain partial content.
    pub fn format<W>(&self, out: W) -> Result<(), Error>
    where
        W: Write,
    {
        let mut out = Renderer::new(out);
        let mut empty_lines = EmptyLines::default();

        // Lower the AST into a series of formatter tokens wrote to `out`.
        format_node(self.t.root_node(), &self.input, &mut empty_lines, &mut out)?;

        out.flush()?;

        Ok(())
    }
}

/// Return the content in `input` for `node.`
fn get_str<'a>(node: &Node<'_>, input: &'a str) -> &'a str {
    &input[node.byte_range()]
}

#[cfg(test)]
mod tests {
    use insta::glob;

    use super::*;

    /// Execute the formatter against each file in the TLA spec corpus, and
    /// compare the output against a reference copy.
    ///
    /// If this test fails, use "cargo insta review" to inspect any output
    /// changes.
    #[test]
    fn test_corpus() {
        use std::fs;

        glob!("../", "tests/corpus/*.tla", |path| {
            let input = fs::read_to_string(path).expect("read test corpus file");
            assert_rewrite!(&input);
        });
    }

    /// A test where the second conj item uses a tab to position the bullet.
    /// With tabs=4 these causes the bullets to align, but a conj_list is not
    /// emitted in the AST.
    ///
    /// ```text
    ///
    ///             ---- MODULE bananas ----
    ///             Op == /\ A = 1
    /// tab ->          /\ B = 2
    ///             ====
    ///
    /// ```
    ///
    /// Emits the following AST nodes:
    ///
    ///     operator_definition
    ///     name: identifier
    ///     def_eq
    ///     definition: bound_infix_op
    ///         lhs: conj_list
    ///         conj_item
    ///             bullet_conj
    ///             bound_infix_op
    ///             lhs: identifier_ref
    ///             symbol: eq
    ///             rhs: nat_number
    ///         symbol: land
    ///         rhs: bound_infix_op
    ///         lhs: identifier_ref
    ///         symbol: eq
    ///         rhs: nat_number
    ///
    /// Note the definition is for bounded_infix_op, with a list of 1 item (the
    /// lhs) with an op of "land" for the second item.
    ///
    /// ```text
    ///
    ///             ---- MODULE Bananas ------
    ///             Op == /\ A = 1
    ///                 /\ B = 2
    ///             =====
    ///
    /// ```
    ///
    /// Compared to the above, that uses only whitespace for alignment:
    ///
    ///     operator_definition
    ///     name: identifier
    ///     def_eq
    ///     definition: conj_list
    ///         conj_item
    ///         bullet_conj
    ///         bound_infix_op
    ///             lhs: identifier_ref
    ///             symbol: eq
    ///             rhs: nat_number
    ///         conj_item
    ///         bullet_conj
    ///         bound_infix_op
    ///             lhs: identifier_ref
    ///             symbol: eq
    ///             rhs: nat_number
    ///
    /// Where the definition node is correctly labelled as a "conj_list" kind,
    /// with two conj_item nodes.
    #[test]
    fn test_mixed_tabs_spaces() {
        assert_rewrite!(
            "\
---- MODULE Bananas ------
X == /\\ x = 4
\t /\\ y = 2
====="
        );
    }
}
