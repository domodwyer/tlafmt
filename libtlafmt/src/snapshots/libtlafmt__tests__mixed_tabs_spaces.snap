---
source: libtlafmt/src/lib.rs
expression: output
---
-------------------------------- MODULE Bananas --------------------------------
X ==
    /\ x = 4
    /\ y = 2
================================================================================
